app/gui/view/graph-editor/src/new_node_position.rs

```diff
@@ -0,0 +1,199 @@
+        let alignment_area_min_y = node_bounding_box.bottom() - bottom.value();
+        let alignment_area = BoundingBox::from_corners(
+            Vector2(alignment_area_min_x, alignment_area_min_y),
+            Vector2(alignment_area_max_x, alignment_area_max_y),
+        );
+        alignment_area.contains(proposed_position)
+    });
+    match alignment_node {
+        Some(node) => under(graph_editor, node.id()),
+        None => proposed_position,
+    }
+}
+
+/// Return a position for a newly created node. Returns a position closely below the `node_id` node
+/// if the position is available, or a first available point on a ray extending to the left of that
+/// position.
+///
+/// Availability of a position is defined in the docs of [`on_ray`].
+pub fn under(graph_editor: &GraphEditorModel, node_above: NodeId) -> Vector2 {
+    let above_pos = graph_editor.node_position(node_above);
+    let y_gap = graph_editor.frp.default_y_gap_between_nodes.value();
+    let y_offset = y_gap + node::HEIGHT;
+    let starting_point = above_pos - Vector2(0.0, y_offset);
+    let direction = Vector2(-1.0, 0.0);
+    on_ray(graph_editor, starting_point, direction).unwrap()
+}
+
+/// Return a position for a newly created node. Return the first available position on a ray
+/// extending from `starting_point` in the `direction`, or [`None`] if the magnitude of each
+/// coordinate of `direction` is smaller than [`f32::EPSILON`].
+///
+/// ## Available position
+///
+/// An available position is a position such that a newly created [`Node`] with origin at this
+/// position does not overlap existing nodes. A node is said to overlap another node if an "overlap
+/// area" (represented by the dashed border in the picture below) around the former intersects with
+/// the bounding box of the latter. A newly created node is assumed to have a fixed size
+/// (represented by the solid border in the picture below). The captions in the picture below are
+/// used as variables in the code.
+/// ```text
+/// ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
+/// ┆                 ▲                    ┆
+/// ┆           y_gap │                    ┆
+/// ┆                 ▼                    ┆
+/// ┆       ┌──────────────────────┐       ┆
+/// ┆       │                   ▲  │       ┆
+/// ┆       │      node::HEIGHT │  │       ┆
+/// ┆       │                   │  │       ┆
+/// ┆       │                   │  │       ┆
+/// ┆ x_gap │  min_spacing      │  │ x_gap ┆
+/// ┆ ◀───▶ │◀──────────────────+─▶│ ◀───▶ ┆
+/// ┆       │                   │  │       ┆
+/// ┆       │                   ▼  │       ┆
+/// ┆       └──────────────────────┘       ┆
+/// ┆                 ▲                    ┆
+/// ┆           y_gap │                    ┆
+/// ┆                 ▼                    ┆
+/// └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
+/// ```
+pub fn on_ray(
+    graph_editor: &GraphEditorModel,
+    starting_point: Vector2,
+    direction: Vector2,
+) -> Option<Vector2> {
```

Comment:
Fast question - this function name suggests, that the ray can be in any direction - does it really work for ray in any direction or only for horizontal and vertical rays?